### DISCUSSION TITLE
What is the functionality of `onBecomeObserved` when `property` (2nd arg) is omitted?

### DISCUSSION THREAD

**Discussion by FirstWhack** [mobxjs/mobx#3355] [Q&A] [answered]
The [relevant documentation](https://mobx.js.org/lazy-observables.html) doesn't seem to explain this, though we can see from the [relevant code](https://github.com/mobxjs/mobx/blob/main/packages/mobx/src/api/become-observed.ts#L56-L57) that the `listener` can actually be passed as the second argument omitting `property` entirely. 

I am wondering what is the functionality of using `onBecomeObserved` like so:

```typescript
    onBecomeObserved(this, () => console.log('became observed');
    onBecomeUnobserved(this, () => console.log('no longer observed');
```

**Comment by urugator** [comment, marked as answer]
Depends on what `this` is. If observable object it will probably throw:
https://github.com/mobxjs/mobx/blob/d233eba03fe977d95589bb3d998bbc9a8ffa776d/packages/mobx/src/types/type-utils.ts#L42
But it can be `observable.box(5)`, `observable.computed(() => {})`, `createAtom()`, `observable([])`, `observable(new Set())`, `observable(new Map())`.
In case of `observable(new Map())` it reacts to subscriptions related only to the keys (addition/removal of key).

**Reply by FirstWhack** [reply]
@urugator  great insights.

Still not sure I understand as we usually pass a Class to the first argument (and that class is not observable).

Am I correct in assuming you can actually pass a reference to the class property (or in general a reference to any observable) if omitting the second argument?

E.g. 
```typescript
class MyClass {
  myProperty = true;
  constructor() {
    makeObservable(this, {
      myProperty: observable
    };
    onBecomeObserved(this.myProperty, () => console.log('myProperty became observed'));
  }
}

```

**Reply by urugator** [reply]
You can pass reference to any observable (not just object), but `this.myProperty` is not a reference to observable, it's boolean. You can't access the observable backing the property directly, therefore you need `onBecomeObserved(obj, key)`, `isObservableProp(obj, key)` etc... 

Actually you can via `getAtom`, so you can do `onBecomeObserved(getAtom(this, 'myProperty'))`, but lets keep it simple...

**Reply by FirstWhack** [reply]
Oh...... Ohhhhhh!!

Thank you so much. This makes sense to me now.
